### PR TITLE
Update Installer and Deployer

### PR DESCRIPTION
**Pull request by sam-heilbron**
# Description

Update the Deployer and Installer created by this repository, so that they can be used to install Gloo Edge and pass the [Google Cloud Marketplace verification process](https://github.com/GoogleCloudPlatform/marketplace-k8s-app-tools/blob/master/docs/mpdev-references.md)

# Context
Based on: https://github.com/solo-io/gcp-gloo-enterprise/pull/4

## Deployer

# Testing
1. Create a GKE cluster
I had [created a cluster](https://console.cloud.google.com/kubernetes/clusters/details/us-central1-a/marketplace-test-sam/details?project=solo-test-236622) which I will remove once this PR has merged.

2. Apply the Application CRD to that cluster
```
kubectl apply -f "https://raw.githubusercontent.com/GoogleCloudPlatform/marketplace-k8s-app-tools/master/crd/app-crd.yaml"
```

3. Verify
```
make mpdev-verify
```

4. Scan images
```
 trivy image --severity HIGH,CRITICAL quay.io/solo-io/gloo:1.15.14
```
```
2023-10-26T08:48:41.471-0400    INFO    Need to update DB
2023-10-26T08:48:41.471-0400    INFO    DB Repository: ghcr.io/aquasecurity/trivy-db
2023-10-26T08:48:41.471-0400    INFO    Downloading DB...
40.68 MiB / 40.68 MiB [------------------------------------------------] 100.00% 17.11 MiB p/s 2.6s
2023-10-26T08:48:44.865-0400    INFO    Vulnerability scanning is enabled
2023-10-26T08:48:44.865-0400    INFO    Secret scanning is enabled
2023-10-26T08:48:44.865-0400    INFO    If your scanning is slow, please try '--security-checks vuln' to disable secret scanning
2023-10-26T08:48:44.865-0400    INFO    Please see also https://aquasecurity.github.io/trivy/0.30.3/docs/secret/scanning/#recommendation for faster secret detection
2023-10-26T08:48:49.144-0400    INFO    Detected OS: ubuntu
2023-10-26T08:48:49.144-0400    INFO    Detecting Ubuntu vulnerabilities...
2023-10-26T08:48:49.149-0400    INFO    Number of language-specific files: 1
2023-10-26T08:48:49.149-0400    INFO    Detecting gobinary vulnerabilities...

quay.io/solo-io/gloo:1.15.14 (ubuntu 20.04)

Total: 0 (HIGH: 0, CRITICAL: 0)


usr/local/bin/gloo (gobinary)

Total: 1 (HIGH: 1, CRITICAL: 0)

┌────────────────────────┬─────────────────────┬──────────┬───────────────────┬────────────────────────┬───────────────────────────────────────────────────┐
│        Library         │    Vulnerability    │ Severity │ Installed Version │     Fixed Version      │                       Title                       │
├────────────────────────┼─────────────────────┼──────────┼───────────────────┼────────────────────────┼───────────────────────────────────────────────────┤
│ google.golang.org/grpc │ GHSA-m425-mq94-257g │ HIGH     │ v1.40.0           │ 1.56.3, 1.57.1, 1.58.3 │ gRPC-Go HTTP/2 Rapid Reset vulnerability          │
│                        │                     │          │                   │                        │ https://github.com/advisories/GHSA-m425-mq94-257g │
└────────────────────────┴─────────────────────┴──────────┴───────────────────┴────────────────────────┴───────────────────────────────────────────────────┘

```
A new go vulnerability that affects Gloo Edge was published 15 hours ago: https://github.com/advisories/GHSA-m425-mq94-257g.

```
 trivy image --severity HIGH,CRITICAL gcr.io/solo-io-public/gloo/deployer:1.15.14
```
```
2023-10-26T08:57:32.121-0400    INFO    Vulnerability scanning is enabled
2023-10-26T08:57:32.121-0400    INFO    Secret scanning is enabled
2023-10-26T08:57:32.121-0400    INFO    If your scanning is slow, please try '--security-checks vuln' to disable secret scanning
2023-10-26T08:57:32.121-0400    INFO    Please see also https://aquasecurity.github.io/trivy/0.30.3/docs/secret/scanning/#recommendation for faster secret detection
2023-10-26T08:57:43.537-0400    INFO    Detected OS: ubuntu
2023-10-26T08:57:43.537-0400    INFO    Detecting Ubuntu vulnerabilities...
2023-10-26T08:57:43.551-0400    INFO    Number of language-specific files: 4
2023-10-26T08:57:43.551-0400    INFO    Detecting gobinary vulnerabilities...
2023-10-26T08:57:43.560-0400    INFO    Detecting python-pkg vulnerabilities...

gcr.io/solo-io-public/gloo/deployer:1.15.14 (ubuntu 20.04)

Total: 0 (HIGH: 0, CRITICAL: 0)


opt/kubectl/1.26/kubectl (gobinary)

Total: 1 (HIGH: 1, CRITICAL: 0)

┌────────────────────────────────┬───────────────┬──────────┬─────────────────────┬───────────────┬───────────────────────────────────────────┐
│            Library             │ Vulnerability │ Severity │  Installed Version  │ Fixed Version │                   Title                   │
├────────────────────────────────┼───────────────┼──────────┼─────────────────────┼───────────────┼───────────────────────────────────────────┤
│ github.com/docker/distribution │ CVE-2023-2253 │ HIGH     │ v2.8.1+incompatible │ 2.8.2-beta.1  │ DoS from malicious API request            │
│                                │               │          │                     │               │ https://avd.aquasec.com/nvd/cve-2023-2253 │
└────────────────────────────────┴───────────────┴──────────┴─────────────────────┴───────────────┴───────────────────────────────────────────┘

opt/kubectl/1.27/kubectl (gobinary)

Total: 1 (HIGH: 1, CRITICAL: 0)

┌────────────────────────────────┬───────────────┬──────────┬─────────────────────┬───────────────┬───────────────────────────────────────────┐
│            Library             │ Vulnerability │ Severity │  Installed Version  │ Fixed Version │                   Title                   │
├────────────────────────────────┼───────────────┼──────────┼─────────────────────┼───────────────┼───────────────────────────────────────────┤
│ github.com/docker/distribution │ CVE-2023-2253 │ HIGH     │ v2.8.1+incompatible │ 2.8.2-beta.1  │ DoS from malicious API request            │
│                                │               │          │                     │               │ https://avd.aquasec.com/nvd/cve-2023-2253 │
└────────────────────────────────┴───────────────┴──────────┴─────────────────────┴───────────────┴───────────────────────────────────────────┘

usr/bin/helm (gobinary)

Total: 2 (HIGH: 2, CRITICAL: 0)

┌──────────────────────────┬─────────────────────┬──────────┬──────────────────────┬────────────────────────┬───────────────────────────────────────────────────┐
│         Library          │    Vulnerability    │ Severity │  Installed Version   │     Fixed Version      │                       Title                       │
├──────────────────────────┼─────────────────────┼──────────┼──────────────────────┼────────────────────────┼───────────────────────────────────────────────────┤
│ github.com/docker/docker │ CVE-2023-28840      │ HIGH     │ v23.0.1+incompatible │ 20.10.24, 23.0.3       │ Encrypted overlay network may be unauthenticated  │
│                          │                     │          │                      │                        │ https://avd.aquasec.com/nvd/cve-2023-28840        │
├──────────────────────────┼─────────────────────┤          ├──────────────────────┼────────────────────────┼───────────────────────────────────────────────────┤
│ google.golang.org/grpc   │ GHSA-m425-mq94-257g │          │ v1.53.0              │ 1.56.3, 1.57.1, 1.58.3 │ gRPC-Go HTTP/2 Rapid Reset vulnerability          │
│                          │                     │          │                      │                        │ https://github.com/advisories/GHSA-m425-mq94-257g │
└──────────────────────────┴─────────────────────┴──────────┴──────────────────────┴────────────────────────┴───────────────────────────────────────────────────┘
```

# Publishing
There is a command in the Makefile that is used to publish the installer and deployer to GCR:
```
make publish
```

This will override the existing published image, so if you are making changes that you want to verify locally, you should change the published version.

**Currently 1.15.14 is published, since that mirrors the version of Gloo Edge that is installed**

# Next Steps
1. I have created https://github.com/solo-io/solo-projects/issues/5462 to address the vulnerability in Gloo Edge, and @sam-heilbron will pick this up ASAP. The goal will be to resolve this in GE 1.15 first, publish a new open source version that we can use in this repo
2. Update the versions of helm/kube used in the installer (likely just a patch bump)

### CODE REVIEW

File: chart/glooctlinstaller/templates/application.yaml [hunk 1]
```diff
@@ -13,7 +13,7 @@ metadata:
 spec:
   descriptor:
     type: Gloo
-    version: '1.3'
+    version: '1.4'
```
Comment (marcogschmidt) [inline chart/glooctlinstaller/templates/application.yaml]:
```suggestion
    version: '1.15'
```

File: Makefile [hunk 2]
```diff
@@ -35,6 +62,10 @@ test-install:
   --deployer=$(REGISTRY)/$(APP_NAME)/deployer:$(DEPLOYER_IMAGE_VERSION) \
   --parameters='{"name": "test-install", "namespace": "$(TEST_NS)"}'
 
+.PHONY: test-uninstall
+test-uininstall:
+	kubectl delete namespace $(TEST_NS)
+
 # copy all the gloo images into the marketplace repo
 .PHONY: docker-mirror
```
Comment (marcogschmidt) [inline Makefile]: We can remove this since you removed the script

File: chart/glooctlinstaller/templates/job.yaml [hunk 3]
```diff
@@ -6,132 +6,80 @@ metadata:
...
+    global:
+      image:
+        registry: gcr.io/gloo-edge
+        pullPolicy: IfNotPresent
+        extended: false
+      glooRbac:
+        namespaced: true
+    settings:
+      singleNamespace: true
+      create: true
```
Comment (marcogschmidt) [inline chart/glooctlinstaller/templates/job.yaml]: Nit: not needed since it's the default

File: Makefile [hunk 4]
```diff
@@ -3,27 +3,54 @@ REGISTRY := gcr.io/solo-io-public
 APP_NAME := gloo
 DEPLOYER_IMAGE_REPO := $(REGISTRY)/$(APP_NAME)/deployer
 INSTALLER_IMAGE_REPO := $(REGISTRY)/$(APP_NAME)/installer
-DEPLOYER_IMAGE_VERSION := 1.3
+DEPLOYER_IMAGE_VERSION := 1.4.0
```
Comment (marcogschmidt) [inline Makefile]: I think this must match the app version

File: chart/glooctlinstaller/templates/job.yaml [hunk 5]
```diff
@@ -6,132 +6,80 @@ metadata:
...
-          registry: {{ .Values.glooImages.gloo.registry }}
-          repository: {{ .Values.glooImages.gloo.repository }}
-          tag: {{ .Values.glooImages.gloo.tag }}
-    discovery:
-      deployment:
-        image:
-          registry: {{ .Values.glooImages.discovery.registry }}
-          repository: {{ .Values.glooImages.discovery.repository }}
-          tag: {{ .Values.glooImages.discovery.tag }}
+        replicas: 1
```
Comment (marcogschmidt) [inline chart/glooctlinstaller/templates/job.yaml]: Also the default

File: schema.yaml [hunk 6]
```diff
@@ -11,179 +11,18 @@ properties:
       type: NAMESPACE
   glooctlInstaller.image.full:
     type: string
-    default: gcr.io/solo-io-public/gloo/installer:1.3
+    default: gcr.io/solo-io-public/gloo/installer:1.4.0
```
Comment (marcogschmidt) [inline schema.yaml]: I think this needs to match the app version as well

File: schema_test_values.yaml [hunk 7]
```diff
@@ -1,13 +1,2 @@
 properties:
   # set a default value (need to copy other fields for validation to pass)
-  enterpriseLicense:
```
Comment (marcogschmidt) [inline schema_test_values.yaml:3]: We can just delete this file and delete this line in the dockerfile
```Dockerfile
COPY schema_test_values.yaml /data-test/schema.yaml
```

File: schema.yaml [hunk 8]
```diff
@@ -11,179 +11,18 @@ properties:
       type: NAMESPACE
```
Comment (marcogschmidt) [inline schema.yaml:35]: I am wondering whether we should use the [V2 of the schema API](https://github.com/GoogleCloudPlatform/marketplace-k8s-app-tools/blob/master/docs/schema.md#v2-specification). It has some fields that are required and do not appear in v1. Just being paranoid here about reasons they might reject the listing.

File: chart/glooctlinstaller/templates/application.yaml (hunk 1 above)
Comment (sam-heilbron) [reply chart/glooctlinstaller/templates/application.yaml]: Updated to match the version

File: Makefile (hunk 2 above)
Comment (sam-heilbron) [reply Makefile]: Good call, removed

File: chart/glooctlinstaller/templates/job.yaml (hunk 3 above)
Comment (sam-heilbron) [reply chart/glooctlinstaller/templates/job.yaml]: Removed

File: Makefile (hunk 4 above)
Comment (sam-heilbron) [reply Makefile]: Good call. To reduce that chance of any version mismatch, I just decided to version our deployer/installer to be identical to the Gloo Edge version. What do you think?

File: schema.yaml (hunk 6 above)
Comment (sam-heilbron) [reply schema.yaml]: Updated

File: schema_test_values.yaml (hunk 7 above)
Comment (sam-heilbron) [reply schema_test_values.yaml:3]: Yeah good call, updated and confirmed the verification still passes

File: schema.yaml (hunk 8 above)
Comment (sam-heilbron) [reply schema.yaml:35]: Good idea. I updated the schema to use v2 and everything is working

File: Makefile (hunk 4 above)
Comment (marcogschmidt) [reply Makefile]: I think that's the recommended way of doing things, i.e. use that same version for everything.

**Review by marcogschmidt — APPROVED** [review]

**Comment by marcogschmidt** [comment]
Note that this gets rid of the option to install Gloo Edge Enterprise. We deemed this code to be unnecessary at the moment, so we would rather delete it and make the repo simpler to understand. We can always add that logic back later.

**Pull request merged**